File: backend/src/main/java/io/tingkai/money/dao/StockRecordDao.java
package io.tingkai.money.dao;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.stereotype.Repository;

import io.tingkai.money.entity.StockRecord;

@Repository
public interface StockRecordDao extends JpaRepository<StockRecord, UUID> {

	public List<StockRecord> findByCode(String code);

	public List<StockRecord> findByCodeOrderByDealDate(String code);

	public List<StockRecord> findByCodeAndDealDateBetweenOrderByDealDate(String code, LocalDateTime start, LocalDateTime end);

	public List<StockRecord> findByCodeAndDealDateBeforeOrderByDealDateDesc(String code, LocalDateTime date, Pageable pageable);

	public Optional<StockRecord> findByCodeAndDealDate(String code, LocalDateTime dealDate);

	public Optional<StockRecord> findFirstByCodeOrderByDealDateDesc(String code);

	@Query(value = "SELECT DISTINCT sr.code FROM stock_record sr", nativeQuery = true)
	public List<String> findDistinctCode();
}

File: python/src/dao/StockRecordDao.py
from datetime import datetime
from typing import List
from uuid import UUID

from entity.StockRecord import StockRecord, db

def findById(id: UUID) -> StockRecord:
    return db.session.query(StockRecord).filter(StockRecord.id == id).first()

def findByCode(code: str) -> List[StockRecord]:
    return db.session.query(StockRecord).filter(StockRecord.code == code).order_by(StockRecord.deal_date).all()

def findByCodeAndDealDate(code: str, dealDate: datetime) -> StockRecord:
    return db.session.query(StockRecord).filter(StockRecord.code == code).filter(StockRecord.deal_date == dealDate).first()

def insert(entity: StockRecord) -> None:
    db.session.add(entity)
    commit()

def update(entity: StockRecord) -> None:
    db.session.add(entity)
    commit()

def commit() -> None:
    db.session.commit()
